SVM.thy

```diff
@@ -0,0 +1,455 @@
+\<^marker>\<open>creator Ralitsa Dardjonova\<close>
+
+theory SVM
+  imports "RLM_stable" "hinge_loss"
+begin
+
+paragraph \<open>Summary\<close>
+text \<open>In this theory we define hinge loss and proove some lemmas 
+that show it can be used as loss function for lipschitz convex learning\<close>
+
+paragraph \<open>Main definitions\<close>
+text \<open>
+\<^item> \<open>soft_svm\<close> the function that is going to be optimized in the soft-SVM optimization problem
+with respect to (w,\<xi>) - respectively the weights and the slack variables.
+             k*|w|^2 + (1/n)\<Sum> \<xi>_i
+\<^item> \<open>argmin_soft_svm\<close> definition of the soft-SVM opimization problem:
+             min_(w,\<xi>) soft_svm, 
+       s.t. \<forall>i. y_i * ( w \<bullet> x_i ) >= 1 - \<xi>_i 
+       and      \<xi>_i >= 0
+\<close>
+
+paragraph \<open>Main Theorems\<close>
+text \<open>
+\<^item> \<open>ksi_are_margins\<close> We prove that the slack variables \<xi> correspond 
+to the distance from a datapoint on the wrong side of the margin to the margin.
+If a datapoint is on the correct side, we have \<xi>_i is 0.
+\<^item> \<open>soft_svm_is_RLM\<close> We show that the soft-SVM optimization problem can be considered
+as a regularized loss minimization problem, in this case ridge regression.
+\<close>
+
+text \<open>We define a locale for learning with hinge loss function.
+It can be used to define soft-SVM optimization problem.\<close>
+locale learning_with_hinge_loss =
+  fixes H :: "'b::euclidean_space set"
+    and D :: "('b \<times> real) pmf"
+    and n :: "nat"
+    and k :: "real"
+  fixes \<rho> :: "real"
+  assumes nnH: "H \<noteq> {}" 
+    and  convH: "convex H"
+    and n_pos: "n>0"
+    and k_pos : "k>0"
+    and rho_pos: "\<rho> > 0"
+    and S_in_D: "S \<in> Samples n D" 
```

Comment:
why is `S` not fixed in the locale as it was in your earlier version?
S seems to be a free variable in `S_in_D`, that seems odd. 
The fact is used in `soft_svm_is_RLM`. I guess `S_in_D` needs to be removed from the locale and put as a assumption in 15.5.
Why did you choose to remove `S_in_D` from the locale assumptions?

Reply:
I removed the assumption, because for claim 15.7 where we take the expectation of the argmin, we change the S, so the definition of the soft svm, shouldn't be fixed. Or maybe I am wrong, there is a workaround or better practice.

Reply:
I see. when you remove `S` from the locale, you also have to remove `S_in_D` from the locale, isn't it?